lib/chef/property.rb

```diff
@@ -518,6 +518,18 @@ def emit_dsl
       # be using the existing getter/setter to manipulate it instead.
       return if !instance_variable_name
 
+      # If the property name is already an existing Ruby method, this could be bad.
+      # For example, the resource may implement #hash which is used internally by
+      # Ruby when adding the object to other data structures, like Hashes.
+      # Let's warn the user that this is bad. Avoid warnings on the core
+      # Chef::Resource class, where properties like `name` may be implemented
+      # and are safe.
+      if declared_in.respond_to?(name) && declared_in != Chef::Resource
+        resource_name = declared_in.respond_to?(:resource_name) ? declared_in.resource_name : declared_in
+        Chef.deprecated(:property_name_collision, "Property `#{name}` of resource `#{resource_name}` is already an existing Ruby method. " \
+          "This could be dangerous and cause unexpected behavior. This will raise an exception in Chef 13.")
```

Comment:
Might want to just say:

"Property `#{name}` of resource `#{resource_name}` overwrites an existing method.  Please use a different property name.  This will raise an exception in Chef 13."

- it doesn't have to be a ruby method, overwriting methods that are on Chef::Resource itself are bad as well (a problem that we have when we add new methods there)
- you burned a lot of words explaining it could be bad, and i don't think that sentence adds any value
- you didn't actually tell them what to do (rename the property)